libdrgn/python/helpers.c

```diff
@@ -249,3 +250,267 @@ PyObject *drgnpy_linux_helper_pgtable_l5_enabled(PyObject *self, PyObject *args,
+		Py_DECREF(index);
+		return NULL;
+	}
+	struct drgn_error* err = drgn_object_copy(&node->obj, &entry->node);
+	if (err) {
+		Py_DECREF(index);
+		Py_DECREF(node);
+		return set_drgn_error(err);
+	}
+	PyObject *ret = PyTuple_Pack(2, index, node);
+	if (!ret) {
+		Py_DECREF(index);
+		Py_DECREF(node);
+		return NULL;
+	}
+	return ret;
+}
+
+static PyObject *idr_for_each_next(GenericIterator *self)
+{
+	struct idr_iter *iterator = (struct idr_iter *)self->iterator;
+	struct radix_tree_iter_entry *entry;
+	struct drgn_error *err = idr_iter_next(iterator, &entry);
+	if (err)
+		return set_drgn_error(err);
+	if (!entry)
+		return NULL;
+	return idr_iter_entry_wrap(entry, self->prog);
+}
+
+GenericIterator *
+drgnpy_linux_helper_idr_for_each(PyObject *self, PyObject *args, PyObject *kwds)
+{
+	static char *keywords[] = {"idr", NULL};
+	struct drgn_error *err;
+	DrgnObject *idr;
+	if (!PyArg_ParseTupleAndKeywords(args, kwds, "O!:idr_for_each",
+					 keywords, &DrgnObject_type, &idr))
+		return NULL;
+
+	GenericIterator *iterator =
+		(GenericIterator *)GenericIterator_type.tp_alloc(
+			&GenericIterator_type, 0);
+	if (!iterator)
+		return NULL;
+	iterator->prog = NULL;
+	iterator->next = idr_for_each_next;
+	iterator->iterator_deinit = (void (*)(void *))idr_iter_deinit;
+	iterator->iterator = malloc(sizeof(struct idr_iter));
+	if (!iterator->iterator) {
+		Py_DECREF(iterator);
+		return set_drgn_error(&drgn_enomem);
+	}
+	err = idr_iter_init(iterator->iterator, &idr->obj);
+	if (err) {
+		Py_DECREF(iterator);
+		return set_drgn_error(err);
+	}
+	iterator->prog = DrgnObject_prog(idr);
+	Py_INCREF(iterator->prog);
+	return iterator;
+}
+
+static PyObject *radix_tree_for_each_next(GenericIterator *self)
+{
+	struct radix_tree_iter *iterator =
+		(struct radix_tree_iter *)self->iterator;
+	struct radix_tree_iter_entry *entry;
+	struct drgn_error *err = radix_tree_iter_next(iterator, &entry);
+	if (err)
+		return set_drgn_error(err);
+	if (!entry)
+		return NULL;
+	return idr_iter_entry_wrap(entry, self->prog);
+}
+
+GenericIterator *drgnpy_linux_helper_radix_tree_for_each(PyObject *self,
+							 PyObject *args,
+							 PyObject *kwds)
+{
+	static char *keywords[] = {"root", NULL};
+	struct drgn_error *err;
+	DrgnObject *root;
+	if (!PyArg_ParseTupleAndKeywords(args, kwds, "O!:radix_tree_for_each",
+					 keywords, &DrgnObject_type, &root))
+		return NULL;
+
+	GenericIterator *iterator =
+		(GenericIterator *)GenericIterator_type.tp_alloc(
+			&GenericIterator_type, 0);
+	if (!iterator)
+		return NULL;
+	iterator->prog = NULL;
+	iterator->next = radix_tree_for_each_next;
+	iterator->iterator_deinit = (void (*)(void *))radix_tree_iter_deinit;
+	iterator->iterator = malloc(sizeof(struct radix_tree_iter));
+	if (!iterator->iterator) {
+		Py_DECREF(iterator);
+		return set_drgn_error(&drgn_enomem);
+	}
+	struct drgn_object root_ptr;
+	drgn_object_init(&root_ptr, &DrgnObject_prog(root)->prog);
+	err = drgn_object_address_of(&root_ptr, &root->obj);
```

Comment:
What is this for? The caller should be passing in a `struct radix_tree_root *`, although we might have some callers that are getting away with `struct radix_tree_root` because the Python API is more permissive about `.` vs `->`.